File: frontend/src/types/shipping.ts
/**
 * TypeScript contracts for the multi-carrier shipping integration.
 *
 * These mirror the backend Pydantic schemas in
 * ``backend/app/schemas/shipping.py``. SECURITY: a carrier account's API key /
 * webhook secret are WRITE-ONLY -- they are sent on create/update but the read
 * shape only ever exposes ``api_key_last4`` + ``has_webhook_secret``. The UI
 * must never request, store, or render a full key.
 */

export type CarrierProvider = 'easypost' | 'zenkraft';
export type CarrierEnvironment = 'production' | 'test';
export type ShipMode = 'parcel' | 'freight';

// ---------------------------------------------------------------------------
// Carrier account (credentials) admin CRUD.
// ---------------------------------------------------------------------------

/** Read shape for a carrier account. NEVER carries the plaintext key/secret. */
export interface CarrierAccount {
  id: number;
  name: string;
  provider: string;
  environment?: string | null;
  is_active?: boolean | null;
  is_default?: boolean | null;
  /** Carrier-ref KEYS only (no values), e.g. ["fedex", "ups"]. */
  carrier_refs: string[];
  /** Masked tail of the stored key, e.g. "a1b2". Null if undecryptable. */
  api_key_last4?: string | null;
  has_webhook_secret: boolean;
  created_at?: string | null;
}

/** Create payload. ``api_key`` / ``webhook_secret`` are write-only secrets. */
export interface CarrierAccountCreate {
  name: string;
  provider: string;
  environment?: string;
  api_key: string;
  webhook_secret?: string | null;
  carrier_refs?: Record<string, string>;
  is_active?: boolean;
  is_default?: boolean;
}

/** Patch payload. Sending ``api_key`` / ``webhook_secret`` rotates the secret. */
export interface CarrierAccountUpdate {
  name?: string;
  environment?: string;
  api_key?: string;
  webhook_secret?: string | null;
  carrier_refs?: Record<string, string>;
  is_active?: boolean;
  is_default?: boolean;
}

export interface CarrierConnectionTestResult {
  ok: boolean;
  provider: string;
  message?: string | null;
}

// ---------------------------------------------------------------------------
// Company shipping profile (ship-from origin + egress kill switch).
// ---------------------------------------------------------------------------

export interface CompanyShippingProfileBase {
  ship_from_name?: string | null;
  ship_from_company?: string | null;
  ship_from_phone?: string | null;
  ship_from_email?: string | null;
  ship_from_street1?: string | null;
  ship_from_street2?: string | null;
  ship_from_city?: string | null;
  ship_from_state?: string | null;
  ship_from_zip?: string | null;
  ship_from_country?: string | null;
  default_package_weight_lbs?: number | string | null;
  default_package_length_in?: number | string | null;
  default_package_width_in?: number | string | null;
  default_package_height_in?: number | string | null;
}

export interface CompanyShippingProfile extends CompanyShippingProfileBase {
  id: number;
  /** Customer-data egress kill switch. Defaults OFF until CUI/DoD sign-off. */
  allow_carrier_egress: boolean;
  created_at?: string | null;
}

export interface CompanyShippingProfileUpdate extends CompanyShippingProfileBase {
  /** Omit to leave unchanged; the egress kill switch is audited when flipped. */
  allow_carrier_egress?: boolean;
}

// ---------------------------------------------------------------------------
// Address validation.
// ---------------------------------------------------------------------------

/** A label-grade postal address at the API boundary. */
export interface ShippingAddress {
  name?: string | null;
  company?: string | null;
  phone?: string | null;
  email?: string | null;
  street1: string;
  street2?: string | null;
  city: string;
  state: string;
  zip: string;
  country?: string;
  residential?: boolean | null;
}

export interface AddressValidationRequest {
  address: ShippingAddress;
}

export interface AddressValidationResult {
  is_valid: boolean;
  normalized: ShippingAddress;
  messages: string[];
  deliverability?: string | null;
}

// ---------------------------------------------------------------------------
// Packages (parcels + pallets) and rate-shop.
// ---------------------------------------------------------------------------

export interface ParcelInput {
  length_in: number | string;
  width_in: number | string;
  height_in: number | string;
  weight_lbs: number | string;
}

export interface PalletInput {
  length_in: number | string;
  width_in: number | string;
  height_in: number | string;
  weight_lbs: number | string;
  freight_class?: string | null;
  nmfc?: string | null;
  stackable?: boolean;
}

export interface RateShopRequest {
  carrier_account_id?: number | null;
  ship_from?: ShippingAddress | null;
  ship_to?: ShippingAddress | null;
  parcels?: ParcelInput[];
  pallets?: PalletInput[];
}

export interface RateQuote {
  /** Persisted ShipmentRateQuote id (null for transient). */
  id?: number | null;
  provider_rate_id: string;
  carrier: string;
  service_code?: string | null;
  service_name?: string | null;
  mode: ShipMode;
  amount: number | string;
  currency: string;
  est_delivery_days?: number | null;
  est_delivery_date?: string | null;
  is_selected: boolean;
}

// ---------------------------------------------------------------------------
// Label / BOL purchase, pickups, void/refund.
// ---------------------------------------------------------------------------

export interface BuyLabelRequest {
  rate_id: string;
  carrier_account_id?: number | null;
}

export interface BuyBolRequest {
  rate_id: string;
  carrier_account_id?: number | null;
}

export interface BuyLabelResult {
  shipment_id: number;
  shipment_number: string;
  carrier?: string | null;
  service_code?: string | null;
  tracking_number?: string | null;
  actual_cost?: number | string | null;
  cost_currency?: string | null;
  label_document_id?: number | null;
  label_purchased_at?: string | null;
  already_purchased: boolean;
}

export interface BuyBolResult {
  shipment_id: number;
  shipment_number: string;
  carrier?: string | null;
  bol_number?: string | null;
  pro_number?: string | null;
  actual_cost?: number | string | null;
  cost_currency?: string | null;
  bol_document_id?: number | null;
  label_purchased_at?: string | null;
  already_purchased: boolean;
}

export interface SchedulePickupRequest {
  /** ISO date, e.g. 2026-06-10. */
  pickup_date: string;
  /** ISO datetime for the earliest pickup window. */
  window_start: string;
  /** ISO datetime for the latest pickup window. */
  window_end: string;
  carrier_account_id?: number | null;
}

export interface SchedulePickupResult {
  provider_pickup_id: string;
  confirmation_number?: string | null;
  scheduled_date?: string | null;
  window_start?: string | null;
  window_end?: string | null;
  status?: string | null;
}

export interface VoidRefundResult {
  shipment_id: number;
  voided_at?: string | null;
  refund_status?: string | null;
  message?: string | null;
}

// ---------------------------------------------------------------------------
// Tracking.
// ---------------------------------------------------------------------------

export interface TrackingEvent {
  id?: number | null;
  status?: string | null;
  status_detail?: string | null;
  occurred_at?: string | null;
  location?: string | null;
  message?: string | null;
  source?: string | null;
  provider_event_id?: string | null;
  created_at?: string | null;
}

export interface ShipmentTracking {
  shipment_id: number;
  shipment_number: string;
  tracking_number?: string | null;
  tracking_status?: string | null;
  tracking_status_detail?: string | null;
  last_tracking_sync_at?: string | null;
  actual_delivery?: string | null;
  events: TrackingEvent[];
}

File: backend/app/schemas/shipping.py
"""Pydantic v2 request/response contracts for the shipping domain.

Covers the legacy manual-shipment shapes (moved verbatim out of
``app/api/endpoints/shipping.py``) PLUS the multi-carrier integration shapes:
carrier-account / shipping-profile admin CRUD, address validation, rate-shop,
buy-label / buy-bol, schedule-pickup, packages, and tracking.

MONEY is always ``Decimal`` (the ORM columns are ``Numeric(12, 2)``); physical
dimensions/weights are ``Decimal`` too. SECRETS are never accepted back nor
returned: a carrier account's API key is write-only on create/update and exposed
only as ``api_key_last4`` on read.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.base import UTCModel

# ---------------------------------------------------------------------------
# Legacy manual-shipment shapes (moved verbatim from the endpoint module).
# ---------------------------------------------------------------------------


class ShipmentCreate(BaseModel):
    work_order_id: int
    ship_to_name: Optional[str] = None
    ship_to_address: Optional[str] = None
    ship_to_city: Optional[str] = None
    ship_to_state: Optional[str] = None
    ship_to_zip: Optional[str] = None
    carrier: Optional[str] = None
    service_type: Optional[str] = None
    quantity_shipped: float
    weight_lbs: Optional[float] = None
    num_packages: int = 1
    packing_notes: Optional[str] = None
    cert_of_conformance: bool = False


class ShipmentUpdate(BaseModel):
    carrier: Optional[str] = None
    service_type: Optional[str] = None
    tracking_number: Optional[str] = None
    ship_date: Optional[date] = None
    estimated_delivery: Optional[date] = None
    status: Optional[str] = None


class ShipmentResponse(UTCModel):
    id: int
    shipment_number: str
    work_order_id: int
    work_order_number: Optional[str] = None
    customer_name: Optional[str] = None
    part_number: Optional[str] = None
    status: str
    ship_to_name: Optional[str] = None
    carrier: Optional[str] = None
    tracking_number: Optional[str] = None
    quantity_shipped: float
    ship_date: Optional[date] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True, use_enum_values=True)


# ---------------------------------------------------------------------------
# Carrier account admin CRUD.
# ---------------------------------------------------------------------------


class CarrierAccountCreate(BaseModel):
    """Create a per-company carrier-aggregator credential.

    ``api_key`` / ``webhook_secret`` are write-only secrets -- they are Fernet
    encrypted before storage and NEVER returned (the response exposes only
    ``api_key_last4``).
    """

    name: str = Field(..., max_length=120)
    provider: str = Field(..., max_length=50, description='"easypost" | "zenkraft"')
    environment: str = Field("production", max_length=20, description='"production" | "test"')
    api_key: str = Field(..., min_length=1, description="Write-only; encrypted at rest, never returned.")
    webhook_secret: Optional[str] = Field(None, description="Write-only; encrypted at rest, never returned.")
    carrier_refs: Dict[str, str] = Field(
        default_factory=dict,
        description='Opaque bring-your-own-carrier account refs, e.g. {"fedex": "...", "ups": "..."}.',
    )
    is_active: bool = True
    is_default: bool = False


class CarrierAccountUpdate(BaseModel):
    """Patch a carrier account. Omitted fields are left unchanged.

    Sending ``api_key`` / ``webhook_secret`` rotates the stored secret; omitting
    them keeps the existing one.
    """

    name: Optional[str] = Field(None, max_length=120)
    environment: Optional[str] = Field(None, max_length=20)
    api_key: Optional[str] = Field(None, min_length=1, description="Write-only; rotates the stored key when present.")
    webhook_secret: Optional[str] = Field(None, description="Write-only; rotates the stored secret when present.")
    carrier_refs: Optional[Dict[str, str]] = None
    is_active: Optional[bool] = None
    is_default: Optional[bool] = None


class CarrierAccountResponse(UTCModel):
    """Carrier account read shape. NEVER exposes the plaintext key/secret."""

    id: int
    name: str
    provider: str
    environment: Optional[str] = None
    is_active: Optional[bool] = None
    is_default: Optional[bool] = None
    carrier_refs: List[str] = Field(default_factory=list, description="Carrier-ref KEYS only (no values).")
    api_key_last4: Optional[str] = None
    has_webhook_secret: bool = False
    created_at: Optional[datetime] = None


class CarrierConnectionTestResponse(BaseModel):
    """Outcome of a credential-only ``test-connection`` (transmits NO customer data)."""

    ok: bool
    provider: str
    message: Optional[str] = None


# ---------------------------------------------------------------------------
# Company shipping profile (ship-from origin + egress kill switch).
# ---------------------------------------------------------------------------


class CompanyShippingProfileBase(UTCModel):
    ship_from_name: Optional[str] = None
    ship_from_company: Optional[str] = None
    ship_from_phone: Optional[str] = None
    ship_from_email: Optional[str] = None
    ship_from_street1: Optional[str] = None
    ship_from_street2: Optional[str] = None
    ship_from_city: Optional[str] = None
    ship_from_state: Optional[str] = None
    ship_from_zip: Optional[str] = None
    ship_from_country: Optional[str] = "US"
    default_package_weight_lbs: Optional[Decimal] = None
    default_package_length_in: Optional[Decimal] = None
    default_package_width_in: Optional[Decimal] = None
    default_package_height_in: Optional[Decimal] = None


class CompanyShippingProfileCreate(CompanyShippingProfileBase):
    # SAFETY: the customer-data egress kill switch DEFAULTS OFF. A human must opt
    # in explicitly (CUI / DoD sign-off) before any outbound carrier call that
    # transmits customer data is permitted.
    allow_carrier_egress: bool = False


class CompanyShippingProfileUpdate(CompanyShippingProfileBase):
    allow_carrier_egress: Optional[bool] = None


class CompanyShippingProfileResponse(CompanyShippingProfileBase):
    id: int
    allow_carrier_egress: bool
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Address validation.
# ---------------------------------------------------------------------------


class AddressSchema(BaseModel):
    """A label-grade postal address at the API boundary (maps to CarrierAddress)."""

    name: Optional[str] = None
    company: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    street1: str
    street2: Optional[str] = None
    city: str
    state: str
    zip: str
    country: str = "US"
    residential: Optional[bool] = None


class AddressValidationRequest(BaseModel):
    address: AddressSchema


class AddressValidationResponse(BaseModel):
    is_valid: bool
    normalized: AddressSchema
    messages: List[str] = Field(default_factory=list)
    deliverability: Optional[str] = None


# ---------------------------------------------------------------------------
# Packages (parcels + pallets) and rate-shop.
# ---------------------------------------------------------------------------


class ParcelSchema(BaseModel):
    length_in: Decimal
    width_in: Decimal
    height_in: Decimal
    weight_lbs: Decimal


class PalletSchema(BaseModel):
    length_in: Decimal
    width_in: Decimal
    height_in: Decimal
    weight_lbs: Decimal
    freight_class: Optional[str] = None
    nmfc: Optional[str] = None
    stackable: bool = False


class RateShopRequest(BaseModel):
    """Rate-shop a shipment. ``parcels`` drives parcel rating; ``pallets`` is the
    LTL/freight path (only honored by a freight-capable provider).

    ``carrier_account_id`` selects which configured account to use; omitting it
    uses the company default. ``ship_to`` / ``ship_from`` override the shipment's
    stored / profile addresses when supplied.
    """

    carrier_account_id: Optional[int] = None
    ship_from: Optional[AddressSchema] = None
    ship_to: Optional[AddressSchema] = None
    parcels: List[ParcelSchema] = Field(default_factory=list)
    pallets: List[PalletSchema] = Field(default_factory=list)


class RateQuoteResponse(UTCModel):
    id: Optional[int] = None  # persisted ShipmentRateQuote id (None for transient)
    provider_rate_id: str
    carrier: str
    service_code: Optional[str] = None
    service_name: Optional[str] = None
    mode: Literal["parcel", "freight"]
    amount: Decimal
    currency: str = "USD"
    est_delivery_days: Optional[int] = None
    est_delivery_date: Optional[date] = None
    is_selected: bool = False

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Label / BOL purchase, pickups.
# ---------------------------------------------------------------------------


class BuyLabelRequest(BaseModel):
    """Buy a parcel label for a previously rate-shopped shipment.

    ``rate_id`` is the ``provider_rate_id`` of the chosen quote (the composite
    ``"<shipment_id>:<rate_id>"`` the adapter emits is accepted as-is).
    """

    rate_id: str
    carrier_account_id: Optional[int] = None


class BuyBolRequest(BaseModel):
    """Buy an LTL Bill of Lading for a freight-rated shipment."""

    rate_id: str
    carrier_account_id: Optional[int] = None


class SchedulePickupRequest(BaseModel):
    """Schedule a carrier pickup for an already-purchased shipment."""

    pickup_date: str = Field(..., description="ISO date, e.g. 2026-06-10.")
    window_start: str = Field(..., description="ISO datetime for the earliest pickup window.")
    window_end: str = Field(..., description="ISO datetime for the latest pickup window.")
    carrier_account_id: Optional[int] = None


class SchedulePickupResponse(UTCModel):
    provider_pickup_id: str
    confirmation_number: Optional[str] = None
    scheduled_date: Optional[date] = None
    window_start: Optional[datetime] = None
    window_end: Optional[datetime] = None
    status: Optional[str] = None


class VoidRefundResponse(UTCModel):
    """Result of a void / refund request (a money-moving CANCEL)."""

    shipment_id: int
    voided_at: Optional[datetime] = None
    refund_status: Optional[str] = None
    message: Optional[str] = None


# ---------------------------------------------------------------------------
# Shipment packages (persisted child rows).
# ---------------------------------------------------------------------------


class ShipmentPackageCreate(BaseModel):
    package_type: Optional[str] = Field(None, description='"box" | "pallet"')
    length_in: Optional[Decimal] = None
    width_in: Optional[Decimal] = None
    height_in: Optional[Decimal] = None
    weight_lbs: Optional[Decimal] = None
    freight_class: Optional[str] = None
    nmfc_code: Optional[str] = None
    quantity: int = 1


class ShipmentPackageResponse(UTCModel):
    id: int
    shipment_id: int
    sequence: Optional[int] = None
    package_type: Optional[str] = None
    length_in: Optional[Decimal] = None
    width_in: Optional[Decimal] = None
    height_in: Optional[Decimal] = None
    weight_lbs: Optional[Decimal] = None
    tracking_number: Optional[str] = None
    freight_class: Optional[str] = None
    nmfc_code: Optional[str] = None
    quantity: Optional[int] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Tracking.
# ---------------------------------------------------------------------------


class TrackingEventResponse(UTCModel):
    id: Optional[int] = None
    status: Optional[str] = None
    status_detail: Optional[str] = None
    occurred_at: Optional[datetime] = None
    location: Optional[str] = None
    message: Optional[str] = None
    source: Optional[str] = None
    provider_event_id: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ShipmentTrackingResponse(UTCModel):
    shipment_id: int
    shipment_number: str
    tracking_number: Optional[str] = None
    tracking_status: Optional[str] = None
    tracking_status_detail: Optional[str] = None
    last_tracking_sync_at: Optional[datetime] = None
    actual_delivery: Optional[date] = None
    events: List[TrackingEventResponse] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Label/BOL purchase results (returned by the buy endpoints).
# ---------------------------------------------------------------------------


class BuyLabelResponse(UTCModel):
    shipment_id: int
    shipment_number: str
    carrier: Optional[str] = None
    service_code: Optional[str] = None
    tracking_number: Optional[str] = None
    actual_cost: Optional[Decimal] = None
    cost_currency: Optional[str] = None
    label_document_id: Optional[int] = None
    label_purchased_at: Optional[datetime] = None
    already_purchased: bool = False


class BuyBolResponse(UTCModel):
    shipment_id: int
    shipment_number: str
    carrier: Optional[str] = None
    bol_number: Optional[str] = None
    pro_number: Optional[str] = None
    actual_cost: Optional[Decimal] = None
    cost_currency: Optional[str] = None
    bol_document_id: Optional[int] = None
    label_purchased_at: Optional[datetime] = None
    already_purchased: bool = False


# Extra normalized tracking-event input the service accepts from jobs/webhooks.
TrackingEventInput = Dict[str, Any]
